### PR TITLE
Realize unified Azure Spring token credentials

### PR DESCRIPTION
**Pull request by saragluna**
Create an alternative for #22472 targeting feature/spring-unify-configurations. 

The SDK-supported credential table:https://github.com/Azure/azure-sdk-for-java/wiki/Spring-Credential.

In a Spring Boot application, we will provide a SpringEnvironmentCredential, the Spring version of EnvironmentCredential, which will read credential properties via Spring properties and default environment variables defined in Azure SDK. We'll also delegate all TokenCredential defined in azure-identity for we can't extend the CredentialBuilderBase. In the AzureContextCredentialAutoConfiguration, we will define several SpringCredentialBuilderBase implementations, order by their precedence, and configure the identity options in a bean factory post-processor, then configure all the credentials in a ChainedTokenCredentialBuilder at last.

When an Azure Spring starter user wants to customize their own TokenCredential, he or she can provide a TokenCredential bean directly or inject a SpringCredentialBuilderBase implementation with higher precedence. An IdentityOptions bean could be defined to customize the identity options.

### CODE REVIEW

File: sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/unity/identity/AzureDefaultTokenCredentialAutoConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    @ConditionalOnMissingBean
+    @Bean
+    @Order(SPRING_ENV_CREDENTIAL_ORDER)
+    public SpringEnvironmentCredentialBuilder springEnvironmentCredentialBuilder(AzureSpringConfiguration azureSpringConfiguration) {
+        return new SpringEnvironmentCredentialBuilder().credentialPropertiesProvider(azureSpringConfiguration);
+    }
+
+    @ConditionalOnMissingBean
+    @Bean
+    @Order(SPRING_ENV_CREDENTIAL_ORDER + 100)
```
Comment (moarychan) [inline sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/unity/identity/AzureDefaultTokenCredentialAutoConfiguration.java:45]: Maybe it's a good choice to implement `org.springframework.core.Ordered` to set the order.

File: sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/unity/identity/AzureDefaultTokenCredentialAutoConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    @Order(SPRING_ENV_CREDENTIAL_ORDER + 500)
+    public SpringAzurePowerShellCredentialBuilder azurePowerShellCredentialBuilder() {
+        return new SpringAzurePowerShellCredentialBuilder();
+    }
+
+    @SuppressWarnings("rawtypes")
+    @Bean
+    @ConditionalOnMissingBean
+    // TODO (xiada) dedicated names for the azureTokenCredential
+    public TokenCredential azureTokenCredential(List<SpringCredentialBuilderBase> credentialBuilders) {
```
Comment (stliu) [inline sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/unity/identity/AzureDefaultTokenCredentialAutoConfiguration.java:82]: prefer `ObjectProvider`?

File: sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/unity/identity/AzureSpringConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+/**
+ * Extends the {@link Configuration} to provide Azure Spring related configurations.
+ */
+public class AzureSpringConfiguration extends Configuration implements CredentialPropertiesProvider {
+
+    private final Configuration configuration;
+
+    public AzureSpringConfiguration(AzureProperties azureProperties) {
+        this.configuration = Configuration.getGlobalConfiguration().clone();
+        overridePropertyIfApplicable(Configuration.PROPERTY_AZURE_TENANT_ID, azureProperties.getCredential().getTenantId());
```
Comment (stliu) [inline sdk/spring/azure-spring-cloud-autoconfigure/src/main/java/com/azure/spring/autoconfigure/unity/identity/AzureSpringConfiguration.java:20]: check this util class `org.springframework.boot.context.properties.PropertyMapper`

**Comment by saragluna** [comment]
Closing this one for it's been integrated in https://github.com/Azure/azure-sdk-for-java/pull/23849.

**Pull request closed without merging**